common/src/main/java/com/tc/net/protocol/transport/ClientConnectionEstablisher.java

```diff
@@ -118,63 +114,56 @@ boolean getAllowReconnects() {
    * @throws CommStackMismatchException
    * @throws MaxConnectionsExceededException
    */
-  public NetworkStackID open(Collection<ConnectionInfo> info, ClientMessageTransport cmt, ClientConnectionErrorListener reporter) throws TCTimeoutException, IOException, MaxConnectionsExceededException,
+  public NetworkStackID open(Iterable<InetSocketAddress> serverAddresses, ClientMessageTransport cmt,
+                             ClientConnectionErrorListener reporter) throws TCTimeoutException, IOException, MaxConnectionsExceededException,
       CommStackMismatchException {
     Assert.assertNotNull(cmt);
     Assert.assertNotNull(reporter);
-    Assert.assertNotNull(info);
+    Assert.assertNotNull(serverAddresses);
     synchronized (this.asyncReconnecting) {
-      if (info != null) {
-        connAddressProvider.addAll(info);
-      }
       Assert.eval("Can't call open() while asynch reconnect occurring", !this.asyncReconnecting.get());
       this.allowReconnects.set(true);
-      return connectTryAllOnce(cmt, reporter);
+      this.serverAddresses = serverAddresses;
+      return connectTryAllOnce(serverAddresses, cmt, reporter);
     }
   }
 
-  NetworkStackID connectTryAllOnce(ClientMessageTransport cmt, ClientConnectionErrorListener reporter) throws TCTimeoutException, IOException,
+  NetworkStackID connectTryAllOnce(Iterable<InetSocketAddress> serverAddresses,
+                                   ClientMessageTransport cmt,
+                                   ClientConnectionErrorListener reporter) throws TCTimeoutException, IOException,
       MaxConnectionsExceededException, CommStackMismatchException {
-    final Iterator<ConnectionInfo> addresses = new ArrayList<ConnectionInfo>(this.connAddressProvider).iterator();
     Assert.assertFalse(cmt.isConnected());
-    ConnectionInfo info = null;
+    Iterator<InetSocketAddress> serverAddressIterator = serverAddresses.iterator();
+    InetSocketAddress target = null;
     
-    while (info != null || addresses.hasNext()) {
-      if (info == null) {
-        info = addresses.next();
+    while (target != null || serverAddressIterator.hasNext()) {
+      if (target == null) {
+        target = nextServerAddress(serverAddressIterator);
       }
       try {
-        return cmt.open(info);
+        return cmt.open(target);
       } catch (TransportRedirect redirect) {
-        ConnectionInfo add = new ConnectionInfo(redirect.getHostname(), redirect.getPort());
-        reporter.onError(info, redirect);
-        info = add;
-        if (this.connAddressProvider.add(add)) {
```

Comment:
removing this causes redirects to be transient.  We want redirects to be permanent additions to the list of known servers.

Reply:
this is not a problem If the user has specified all the servers in the stripe, right?

Reply:
true but that isn't going to happen all the time.

Reply:
plus that ability was there before.  we should preserve it.

Reply:
Made some changes to preserve redirects, please have a look